### PR TITLE
Deps: Migrate to ESLint v9.x and FlatConfig

### LINKED ISSUE #948: Migrate to ESLint v9.x and FlatConfig
ESLint v8.x (currently v8.57.0) reached EOL last month. To use later TypeScript versions and ES2024, we should migrate to version 9.x and switch to FlatConfig simultaneously.

Migration guides:
- https://eslint.org/docs/latest/use/migrate-to-9.0.0
- https://eslint.org/docs/latest/use/configure/migration-guide

### PR DESCRIPTION
**Pull request by 64qq**
Closes #948 
<details>
<summary>See the tasklist in the original issue for what this PR does</summary>

> Tasks on my local repo:
> 
> * [x]  Upgrade dependencies
>   
>   * [x]  `eslint` @8.57.0 -> @9.15.0
>   * [x]  `eslint-config-love` @43.1.0 -> @107.0.0
>   * [x]  `typescript-eslint` @6.21.0 -> @8.15.0
> * [x]  Migrate to FlatConfig
>   
>   * [x]  Adapt to dependencies' breaking changes
>     
>     * [x]  Permanent fixes
>       
>       * [x]  Replace `@typescript-eslint/ban-types` with `@typescript-eslint/no-empty-object-type` in directives ([typescript-eslint@8.0.0](https://github.com/typescript-eslint/typescript-eslint/releases/tag/v8.0.0))
>       * [x]  Use `parserOptions.projectService` instead of `parserOptions.project` ([typescript-eslint@8.0.0](https://github.com/typescript-eslint/typescript-eslint/releases/tag/v8.0.0), [blog post](https://typescript-eslint.io/blog/announcing-typescript-eslint-v8/#project-service), [troubleshooting](https://typescript-eslint.io/troubleshooting/typed-linting/#id-like-to-use-tsconfigs-other-than-tsconfigjsons-for-project-service-type-information))
>         
>         * [x]  Remove `tsconfig-eslint.json`
>         * [x]  Add `tsconfig-build.json` to the project `graph-state`
>         * [x]  Set option `allowJs` to `true` in `tsconfig-base.json` to fix the errors like `Parsing error: something.js was not found by the project service`
>         * [x]  Explicitly include the files outside the project to fix TS6307
>       * [x]  [autofix] Apply `@typescript-eslint/consistent-type-exports` to `export *` exports ([typescript-eslint@8.7.0](https://github.com/typescript-eslint/typescript-eslint/releases/tag/v8.7.0))
>       * [x]  [partial autofix] Apply `@typescript-eslint/consistent-indexed-object-style` to mapped types ([typescript-eslint@8.12.1](https://github.com/typescript-eslint/typescript-eslint/releases/tag/v8.12.1))
>       * [x]  [autofix] Apply `@typescript-eslint/no-duplicate-type-constituents` ([love@57.0.0](https://github.com/mightyiam/eslint-config-love/releases/tag/v57.0.0))
>       * [x]  Replace `@typescript-eslint/no-var-requires` with `@typescript-eslint/no-require-imports` in directives ([love@63.0.0](https://github.com/mightyiam/eslint-config-love/releases/tag/v63.0.0))
>       * [x]  [autofix] Apply `@typescript-eslint/no-unnecessary-template-expression` ([love@73.0.0](https://github.com/mightyiam/eslint-config-love/releases/tag/v73.0.0))
>       * [x]  [partial autofix] Apply `@typescript-eslint/prefer-destructuring` ([love@89.0.0](https://github.com/mightyiam/eslint-config-love/releases/tag/v89.0.0))
>         
>         * [x]  Set option `enforceForRenamedProperties` to `false`
>       * [x]  Apply `@typescript-eslint/prefer-for-of` ([love@91.0.0](https://github.com/mightyiam/eslint-config-love/releases/tag/v91.0.0))
>       * [x]  [autofix] Apply `@typescript-eslint/prefer-regexp-exec` ([love@94.0.0](https://github.com/mightyiam/eslint-config-love/releases/tag/v94.0.0))
>       * [x]  [autofix] Apply `@typescript-eslint/prefer-string-starts-ends-with` ([love@95.0.0](https://github.com/mightyiam/eslint-config-love/releases/tag/v95.0.0))
>       * [x]  [autofix] Apply `arrow-body-style` ([love@96.0.0](https://github.com/mightyiam/eslint-config-love/releases/tag/v96.0.0))
>     * [x]  Workarounds
>       
>       * [x]  Set option `allowThrowingUnknown` of `@typescript-eslint/only-throw-error` to `true` ([love@51.0.0, commit 6db9c2b](https://github.com/mightyiam/eslint-config-love/commit/6db9c2ba1e8e95a7e31589fc999ceef60ce48b70#diff-a2a171449d862fe29692ce031981047d7ab755ae7f84c707aef80701b3ea0c80R154))
>       * [x]  Disable `@typescript-eslint/class-methods-use-this` ([love@55.0.0](https://github.com/mightyiam/eslint-config-love/releases/tag/v55.0.0))
>       * [x]  Disable `@typescript-eslint/max-params` ([love@57.0.0](https://github.com/mightyiam/eslint-config-love/releases/tag/v57.0.0))
>       * [x]  ~Disable `@typescript-eslint/no-duplicate-type-constituents` ([love@57.0.0](https://github.com/mightyiam/eslint-config-love/releases/tag/v57.0.0))~
>       * [x]  Disable `@typescript-eslint/init-declarations` ([love@57.0.0](https://github.com/mightyiam/eslint-config-love/releases/tag/v57.0.0))
>       * [x]  Disable `@typescript-eslint/no-empty-function` ([love@57.0.0](https://github.com/mightyiam/eslint-config-love/releases/tag/v57.0.0))
>       * [x]  Disable `@typescript-eslint/no-explicit-any` ([love@58.0.0](https://github.com/mightyiam/eslint-config-love/releases/tag/v58.0.0))
>       * [x]  Disable `@typescript-eslint/no-loop-func` ([love@60.0.0](https://github.com/mightyiam/eslint-config-love/releases/tag/v60.0.0))
>       * [x]  Disable `@typescript-eslint/no-inferrable-types` ([love@62.0.0](https://github.com/mightyiam/eslint-config-love/releases/tag/v62.0.0))
>       * [x]  Disable `@typescript-eslint/no-deprecated` ([love@64.0.0](https://github.com/mightyiam/eslint-config-love/releases/tag/v64.0.0))
>       * [x]  Disable `@typescript-eslint/no-magic-numbers` ([love@65.0.0](https://github.com/mightyiam/eslint-config-love/releases/tag/v65.0.0))
>       * [x]  Disable `@typescript-eslint/no-redundant-type-constituents` ([love@69.0.0](https://github.com/mightyiam/eslint-config-love/releases/tag/v69.0.0))
>       * [x]  Disable `@typescript-eslint/no-unnecessary-condition` ([love@70.0.0](https://github.com/mightyiam/eslint-config-love/releases/tag/v70.0.0))
>       * [x]  ~Disable `@typescript-eslint/no-unnecessary-template-expression` ([love@73.0.0](https://github.com/mightyiam/eslint-config-love/releases/tag/v73.0.0))~
>       * [x]  Disable `@typescript-eslint/no-unnecessary-type-arguments` ([love@74.0.0](https://github.com/mightyiam/eslint-config-love/releases/tag/v74.0.0))
>       * [x]  Disable `@typescript-eslint/no-unnecessary-type-parameters` ([love@75.0.0](https://github.com/mightyiam/eslint-config-love/releases/tag/v75.0.0))
>       * [x]  Disable `@typescript-eslint/no-unsafe-assignment` ([love@76.0.0](https://github.com/mightyiam/eslint-config-love/releases/tag/v76.0.0))
>       * [x]  Disable `@typescript-eslint/no-unsafe-call` ([love@77.0.0](https://github.com/mightyiam/eslint-config-love/releases/tag/v77.0.0))
>       * [x]  Disable `@typescript-eslint/no-unsafe-member-access` ([love@80.0.0](https://github.com/mightyiam/eslint-config-love/releases/tag/v80.0.0))
>       * [x]  Disable `@typescript-eslint/no-unsafe-return` ([love@81.0.0](https://github.com/mightyiam/eslint-config-love/releases/tag/v81.0.0))
>       * [x]  Disable `@typescript-eslint/no-unsafe-unary-minus` ([love@82.0.0](https://github.com/mightyiam/eslint-config-love/releases/tag/v82.0.0))
>       * [x]  ~Disable `@typescript-eslint/prefer-destructuring` ([love@89.0.0](https://github.com/mightyiam/eslint-config-love/releases/tag/v89.0.0))~
>       * [x]  ~Disable `@typescript-eslint/prefer-regexp-exec` ([love@94.0.0](https://github.com/mightyiam/eslint-config-love/releases/tag/v94.0.0))~
>       * [x]  ~Disable `@typescript-eslint/prefer-string-starts-ends-with` ([love@95.0.0](https://github.com/mightyiam/eslint-config-love/releases/tag/v95.0.0))~
>       * [x]  ~Disable `arrow-body-style` ([love@96.0.0](https://github.com/mightyiam/eslint-config-love/releases/tag/v96.0.0))~
>       * [x]  Disable `eslint-comments/require-description` ([love@97.0.0](https://github.com/mightyiam/eslint-config-love/releases/tag/v97.0.0))
>       * [x]  Disable `@typescript-eslint/switch-exhaustiveness-check` ([love@98.0.0](https://github.com/mightyiam/eslint-config-love/releases/tag/v98.0.0))
>       * [x]  Disable `@typescript-eslint/no-unsafe-type-assertion` ([love@101.0.0](https://github.com/mightyiam/eslint-config-love/releases/tag/v101.0.0))
>       * [x]  Disable `@typescript-eslint/use-unknown-in-catch-callback-variable` ([love@103.0.0](https://github.com/mightyiam/eslint-config-love/releases/tag/v103.0.0))
>       * [x]  Disable `@typescript-eslint/require-await` ([love@104.0.0](https://github.com/mightyiam/eslint-config-love/releases/tag/v104.0.0))
>   * [x]  Remove unused eslint-disable directives
> * [x]  Migrate custom rules (https://eslint.org/blog/2023/09/preparing-custom-rules-eslint-v9)
</details>
Strike-through lines indicate that the workaround has been replaced by a permanent fix.

### CODE REVIEW

File: tsconfig-base.json [hunk 1]
```diff
@@ -21,7 +21,7 @@
     "jsx": "react",
     "jsxFactory": "jsx",
 
-    "allowJs": false,
+    "allowJs": true,
```
Comment (jared-hughes) [inline tsconfig-base.json:24]: Hmm

File: eslint.config.mjs [hunk 2]
```diff
@@ -70,6 +70,15 @@ export default tseslint.config(
...
+
+      "@typescript-eslint/prefer-destructuring": [
+        "error",
+        { array: true, object: true },
+        {
+          enforceForRenamedProperties: false,
+          enforceForDeclarationWithTypeAnnotation: false,
+        },
+      ],
     },
```
Comment (jared-hughes) [inline eslint.config.mjs]: There's a few weird situations from this new rule (like `({ left: this.x, top: this.y } = bbox);`) , but I still count it as a net positive.

File: eslint.config.mjs [hunk 3]
```diff
@@ -79,6 +79,35 @@ export default tseslint.config(
...
+      "@typescript-eslint/no-unnecessary-condition": "off",
+      "@typescript-eslint/no-unnecessary-type-arguments": "off",
+      "@typescript-eslint/no-unnecessary-type-parameters": "off",
+      "@typescript-eslint/no-unsafe-assignment": "off",
+      "@typescript-eslint/no-unsafe-call": "off",
+      "@typescript-eslint/no-unsafe-member-access": "off",
+      "@typescript-eslint/no-unsafe-return": "off",
+      "@typescript-eslint/no-unsafe-unary-minus": "off",
+      "eslint-comments/require-description": "off",
+      "@typescript-eslint/switch-exhaustiveness-check": "off",
```
Comment (jared-hughes) [inline eslint.config.mjs:107]: :eyes: `@typescript-eslint/switch-exhaustiveness-check` looks niiice

**Review by jared-hughes — APPROVED** [review]
Great work here! Slightly concerned about `allowJs` becoming `true` (I assume it's needed for the `.mjs` files?), but that shouldn't really be a problem. Rather not fight it.

`@typescript-eslint/prefer-destructuring` is interesting and a net positive, but didn't necessarily need to be enabled.

File: eslint.config.mjs (hunk 2 above)
Comment (64qq) [reply eslint.config.mjs]: I didn't know that `enforceForRenamedProperties` defaults to true in eslint-config-love, so was trying to fix all the errors including those for renamed properties at first. I seem to have forgotten to revert the changes made at that time when I disabled that option later.

**Pull request merged**